components/raftstore/src/store/fsm/apply.rs

```diff
@@ -1210,7 +1213,13 @@ where
 
         if !data.is_empty() {
             if !self.peer.is_witness || !can_witness_skip(entry) {
-                let cmd = util::parse_data_at(data, index, &self.tag);
+                let cmd = match util::parse_raft_cmd_request(data, index, term, &self.tag) {
+                    util::RaftCmd::V1(cmd) => cmd,
+                    util::RaftCmd::V2(simple_write_decoder) => {
+                        assert!(apply_ctx.enable_v2_compatible_learner);
```

Comment:
Should return error?

Reply:
why can't use assert here? And the method returns `ApplyResult`, so I don't know what kind of error to return?

Reply:
An assertion *asserts* things, i.e. things that are implied by the code itself (versus by the user). I understand this flag should be set as true by tiflash, but tiflash doesn't belong to tikv, so you shouldn't rely on the invariant of tiflash to assert things in tikv.

> I don't know what kind of error to return

It is a bit difficult. You can panic instead with sufficient panic message.